README
NAME
    Colloquy::Data - Read Colloquy 1.3 and 1.4 data files

SYNOPSIS
     use Data::Dumper;
     use Colloquy::Data qw(:all);
     
 my $colloquy_datadir = "/home/system/colloquy/data";
     
 #my ($users_hashref,$lists_hashref) = users($colloquy_datadir);
     my ($lists_hashref,$users_hashref) = lists($colloquy_datadir);
     
 print "Users: ".Dumper($users);
     print "Lists: ".Dumper($lists);

DESCRIPTION
    This module munges the users.lua and lists.lua (Colloquy 1.3x) files in
    to executable perl code which is then evaluated. Colloquy 1.4 uses a
    seperate LUA file for each user and list, which are located in the users
    and lists directories in the Colloquy data directory. These files are
    read one by one and evaluated in the same way.

    This module should therefore be used with caution if you cannot
    gaurentee the integrity of the user and list LUA files! The module will
    issue a warning complaining about write group permissions if $^W
    warnings are enabled, and will die if any of the LUA files have world
    writable permissions.

EXPORTS
  users
     my ($users_hashref,$lists_hashref) = users($colloquy_datadir);

    Returns users and lists hash references, in that order.

  lists
     my ($lists_hashref,$users_hashref) = lists($colloquy_datadir);

    Returns lists and users hash references, in that order.

SEE ALSO
    <http://freshmeat.net/projects/colloquy-talker/>

VERSION
    $Id: Data.pm 522 2006-05-29 11:47:31Z nicolaw $

AUTHOR
    Nicola Worthington <nicolaw@cpan.org>

    <http://perlgirl.org.uk>

COPYRIGHT
    Copyright 2005,2006 Nicola Worthington.

    This software is licensed under The Apache Software License, Version
    2.0.

    <http://www.apache.org/licenses/LICENSE-2.0>



Changes
Version 1.15 - 29th May 2006
    + Changed to use Module::Build.

    + Added use of Safe module to prevent code injection
      through the evaluation of the Colloquy data files.

Version 1.14 - 6th February 2006
    + Modified Makefile.PL to be aware of automated testing
      environment variable and default to no to send usage
      information.

Version 1.13 - 12th January 2006
    + Fixed POD so that it actually passes the
      Test::Pod::Coverage unit tests, and fixed
      the unit test so that it actually tests the
      POD coverage since it wasn't before.

Version 1.12 - 12th January 2006
    + Tweaked unit test for 1.4 data to ensure
      that permissions are safe to begin with, and
      then force an insecurity to test the security
      trap.

Version 1.11 - 7th January 2006
    + Added Test::Pod and Test::Pod::Coverage unit
      tests to make more compliant with CPANTS.

Version 1.10 - 30th December 2005
    + Changed license to Apache 2.0
      http://www.apache.org/licenses/LICENSE-2.0

Version 1.9 - 29th December 2005
    + Added Changes file.

    + Added GPL license header comment to various files.

    + Modified Makefile.PL to send anonymous information to
      http://perlgirl.org.uk/lib/usage.cgi if the user
      agrees. This is so I can track which of my modules
      are most heavily used and on what platforms, so I can
      concentrate on maintaning the most popular ones in
      my limited spare time.

Version 1.6 - 13th December 2005
    + Changed code to only read in files and ignore
      directories.

Version 1.5 - 8th December 2005
    + Changed POD.

Version 1.4 - 28th November 2005
    + Added condition to catch known invalid user data files
      in colloquy when only the "return {" line is written
      because of case sensitive cruft in the colloquy code.

Version 1.3 - 12th November 2005
    + Fixed typo.

Version 1.2 - 12th November 2005
    + Updated the code to include carping and to actually
      do the permissions checking ;-)

Version 1.1 - 12th November 2005
    + Adding initial release.



Build.PL
# vim:ts=4:sw=4:tw=78
# $Id$

use strict;
use Module::Build;
use vars qw($build);

$build = Module::Build->new(
		module_name        => 'Colloquy::Data',
		license            => 'open_source',
		create_makefile_pl => 'passthrough',
		create_readme      => 1,
		create_packlist    => 1,
		sign               => 0,

		requires => {
			'Carp'         => 0, # 1.04
			'Safe'         => 2.11, # 2.11
		},

		recommends => {
			'Test'                => 0, # 1.25
			'Test::More'          => 0, # 0.60
			'Test::Pod'           => 0, # 1.20
			'Test::Pod::Coverage' => 0, # 1.06
			'Test::Assertions'    => 0,
		},
	);

$build->create_build_script;


# Send perl and module version information home if we've been given
# permission to do so by a human being - default to not send for automated
# testing environments, of if the user does not respond within 20 seconds.

my $url = $ENV{AUTOMATED_TESTING} ? undef : may_send_version_information();
if ($url) {
	my @resp = ();
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 10;
		my $ua = LWP::UserAgent->new(
				agent => 'Build.PL $Revision$',
				timeout => 9,
				max_size => 500,
			);
		$ua->env_proxy;
		my $response = $ua->get($url);
		if ($response->is_success()) {
			for (split(/\s*\n+\s*/, $response->content())) {
				push @resp, $_ if $_;
			}
		}
		alarm 0;
	};
	print substr($resp[0],0,79) || "Thank you for sending this information.";
	print "\n\n";
}

sub may_send_version_information {
	eval {
		require Config;
		require LWP::UserAgent;
	};
	return undef if $@;

	my $str = sprintf('%s?%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s',
			'http://perlgirl.org.uk/lib/usage.cgi',
			'name',     $build->dist_name(),
			'version',  $build->dist_version(),
			'osname',   $Config::Config{osname},
			'archname', $Config::Config{archname},
			'osver',    $^O,
			'perlver',  $]
		);

	print "\nThank you for downloading ".$build->dist_name()."\n\n";
	print "I would like to find out how many people are using this software,\n";
	print "and on what operating systems and Perl versions. If you have an\n";
	print "internet connection, may I transmit the following information:\n\n";
	print "$str\n\n";

	my $send = 0;
	eval {
		local $SIG{ALRM} = sub { die; };
		alarm 20;
		$send = $build->y_n('Send this anonymous information?','n');
		alarm 0;
	};

	return $send ? $str : undef;
}

1;



Makefile.PL
# vim:ts=4:sw=4:tw=78

require 5.005;
use strict;
use ExtUtils::MakeMaker qw(WriteMakefile);

$|++;

my $name = 'Colloquy::Data';
my $from = 'lib/Colloquy/Data.pm';

# Make the Makefile as per usual
WriteMakefile(
		NAME          => $name,
		VERSION_FROM  => $from,
		ABSTRACT_FROM => $from,
		AUTHOR        => 'Nicola Worthington <nicolaw@cpan.org>',
		PREREQ_PM     => {
				'Carp' => 0,
				'Test::More' => 0,
				'Test::Assertions' => 0,
			},
	);

# Send perl and module version information home if we've been
# given permission to do so by a human being - default to not
# send for automated testing environments, of if the user does
# not respond within 20 seconds.
my $url = $ENV{AUTOMATED_TESTING} ? undef : may_send_version_information();
if ($url) {
	my @resp = ();
	eval {
		local $SIG{ALRM} = sub { exit; };
		alarm 10;
		for (split(/\s*\n+\s*/, LWP::Simple::get($url))) {
			push @resp, $_ if $_;
		}
		alarm 0;
	};
	print substr($resp[0],200) || "Thank you for sending this information.";
	print "\n\n";
}

sub may_send_version_information {
	eval {
		require Term::ReadKey;
		require Config;
		require LWP::Simple;
	};
	return undef if $@;

	my $version = '';
	eval { $version = ExtUtils::MM_Unix::parse_version({},$from); };
	if (!$version && open(FH,'<META.yml')) {
		while (local $_ = <FH>) {
			if (/version:\s*([0-9\.]+)/i) {
				$version = $1;
				last;
			}
		} close(FH);
	}

	my $str = sprintf('%s?%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s',
			'http://perlgirl.org.uk/lib/usage.cgi',
			'name',     $name,
			'version',  $version,
			'osname',   $Config::Config{osname},
			'archname', $Config::Config{archname},
			'osver',    $^O,
			'perlver',  $]
		);

	print "\nThank you for downloading $name\n\n";
	print "I would like to find out how many people are using this software,\n";
	print "and on what operating systems and Perl versions. If you have an\n";
	print "internet connection, may I transmit the following information:\n\n";
	print "$str\n";
	print "\nSend this anonymous information? (Y/n): ";

	my $send = 0;
	Term::ReadKey::ReadMode(4);
	1 while (defined($send = Term::ReadKey::ReadKey(20)) && $send !~ /^[yn]*$/i);
	$send = $send =~ /^y$/i ? 1 : 0;
	print $send ? "y\n\n" : "n\n\n";
	Term::ReadKey::ReadMode(0);

	return $send ? $str : undef;
}



lib/Colloquy/Data.pm
############################################################
#
#   $Id$
#   Colloquy::Data - Read Colloquy 1.3 and 1.4 data files
#
#   Copyright 2005,2006 Nicola Worthington
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
############################################################

package Colloquy::Data;
# vim:ts=4:sw=4:tw=78

use strict;
use Exporter;
use Fcntl ':mode';
use Carp qw(cluck croak);
use Safe;

use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS $DEBUG);
use constant DEFAULT_DATADIR => '/usr/local/colloquy/data';

$VERSION     = '1.15' || sprintf('%d', q$Revision$ =~ /(\d+)/g);
$DEBUG       = $ENV{DEBUG} ? 1 : 0;
@ISA         = qw(Exporter);
@EXPORT      = ();
@EXPORT_OK   = qw(&lists &users &caps &commify);
%EXPORT_TAGS = ( all => \@EXPORT_OK );

sub users {
	return _get_data(shift);
}

sub lists {
	my ($users,$lists) = _get_data(shift);
	return ($lists,$users);
}

sub caps {
	(my $c = $_[0]) =~ s/_/ /g;
	my @c = split(/\b/,$c);
	foreach (@c) { if (/^([a-z])(.*)/) { $_ = uc($1).$2; } }
	return join("",@c);
}

sub commify {
	local $_ = shift;
	s/^\s+|\s+$//g;
	1 while s/^([-+]?\d+)(\d{3})/$1,$2/;
	return $_;
}

sub _munge_user_lua {
	local $_ = shift;
	s/'/\\'/g;
	s/"/'/g; #"'
	s/(\s+[a-z0-9]+\s+=)(\s+['{\d+])/$1>$2/gi;
	s/^return //;
	return $_;
}

sub _munge_list_lua {
	local $_ = shift;
	s/\s+\['(\S+?)'\]\s+=\s+{/ $1 => {/g;
	s/'/\\'/g;
	s/"/'/g; #"'
	s/(\s+[a-z0-9]+\s+=)(\s+['{\d+])/$1>$2/gi;
	s/(\s+members\s+=>\s+)\{(.+?)\}/$1 [ ( $2 ) ]/sgi;
	s/^return //;
	return $_;
}

sub _read_file {
	my $file = shift;
	croak "No such file '$file'\n" unless -e $file;
	croak "'$file' is not a plain file type\n" unless -f _;
	croak "Insufficient permissions to read file '$file'\n" unless -r _;

	my $mode = (stat(_))[2];
	my $group_write = ($mode & S_IWGRP) >> 3;
	my $other_write = $mode & S_IWOTH;

	# Since this module started using Safe to parse the data files,
	# this code is no longer as important as before. It's now only
	# a warning.
#	if ($^W && $group_write) {
#		cluck "WARNING! $file is group writeable. This is potentially insecure!";
#	}
	#if ($other_write) {
	if ($^W && $other_write) {
		#croak "FATAL! $file is world writeable. This insecure file cannot be evaluated!";
		cluck "WARNING! $file is world writeable. This is potentially insecure!";
	}

	if (open(FH,"<$file")) {
		local $/ = undef;
		my $data = <FH>;
		close(FH);
		return $data;
	} else {
		croak "Unable to open file handle FH for file '$file': $!";
		# return undef;
	}
}

sub _get_data {
	my $datadir = shift || DEFAULT_DATADIR;
	my $users_lua = $datadir.'/users'.(-f $datadir.'/users.lua' ? '.lua' : '');
	my $lists_lua = $datadir.'/lists'.(-f $datadir.'/lists.lua' ? '.lua' : '');

	my $users = {};
	croak "Insufficient permissions to read $users_lua\n" unless -r $users_lua;

	my $c = new Safe;
	# Minimum safe opcode set for building data structures lineseq, list and
	# padany needed for perl 5.8.7
	$c->permit_only(qw(rv2sv sassign aelem aelemfast helem anonlist anonhash
				pushmark refgen const undef leaveeval lineseq list padany));

	if (-f $users_lua) {
		my $coderef = _munge_user_lua( '$' . _read_file($users_lua) );
		$users = $c->reval($coderef);
		#eval $coderef;

	} elsif (-d $users_lua) {
		if (opendir(DH,$users_lua)) {
			for my $user (grep(!/^\./,readdir(DH))) {
				next unless -f "$users_lua/$user";
				unless (-r "$users_lua/$user") {
					cluck "Insufficient permissions to read $users_lua/$user";
					next;
				}
				my $coderef = _munge_user_lua( _read_file("$users_lua/$user") );
				if (length($coderef) > 9 && $coderef =~ /^\s*(return )?{.+}\s*$/gsi) {
#				if (length($coderef) > 9 && $coderef =~ /return {.+}/gsi) {
					DUMP('$coderef',$coderef);
					$users->{$user} = $c->reval($coderef);
					DUMP('$users',$users);
					#eval { $users->{$user} = eval $coderef; }
				} else {
					cluck "Caught known Colloquy data file corruption for user $user";
				}
			}
			closedir(DH);
		} else {
			croak "Failed to open file handle DH for directory '$users_lua': $!";
		}
	}

	my $lists = {};
	croak "Insufficient permissions to read $lists_lua\n" unless -r $lists_lua;

	if (-f $lists_lua) {
		my $coderef = _munge_list_lua( '$' . _read_file($lists_lua) );
		$lists = $c->reval($coderef);
		#eval $coderef;

	} elsif (-d $lists_lua) {
		if (opendir(DH,$lists_lua)) {
			for my $list (grep(!/^\./,readdir(DH))) {
				next unless -f "$lists_lua/$list";
				unless (-r "$lists_lua/$list") {
					cluck "Insufficient permissions to read $lists_lua/$list";
					next;
				}
				my $coderef = _munge_list_lua( _read_file("$lists_lua/$list") );
				if (length($coderef) > 9 && $coderef =~ /^\s*(return )?{.+}\s*$/gsi) {
#				if (length($coderef) > 9 && $coderef =~ /return {.+}/gsi) {
					DUMP('$coderef',$coderef);
					$lists->{$list} = $c->reval($coderef);
					DUMP('$lists',$lists);
					#$lists->{$list} = eval $coderef;
				} else {
					cluck "Caught known Colloquy data file corruption for list $list";
				}
			}
			closedir(DH);
		} else {
			croak "Failed to open file handle DH for directory '$lists_lua': $!";
		}
	}

	for my $list (keys %{$lists}) {
		for my $member (@{$lists->{$list}->{members}}) {
			$users->{$member}->{lists} = [] unless exists $users->{$member}->{lists};
			$lists->{$list}->{users} = [] unless exists $lists->{$list}->{users};
			push @{$users->{$member}->{lists}},$list;
			push @{$lists->{$list}->{users}},$member;
		}
	}

	return ($users,$lists);
}

sub TRACE {
	return unless $DEBUG;
	warn(shift());
}

sub DUMP {
	return unless $DEBUG;
	eval {
		require Data::Dumper;
		warn(shift().': '.Data::Dumper::Dumper(shift()));
	}
}

1;

=pod

=head1 NAME

Colloquy::Data - Read Colloquy 1.3 and 1.4 data files

=head1 SYNOPSIS

 use Data::Dumper;
 use Colloquy::Data qw(:all);
 
 my $colloquy_datadir = "/home/system/colloquy/data";
 
 #my ($users_hashref,$lists_hashref) = users($colloquy_datadir);
 my ($lists_hashref,$users_hashref) = lists($colloquy_datadir);
 
 print "Users: ".Dumper($users);
 print "Lists: ".Dumper($lists);

=head1 DESCRIPTION

This module munges the users.lua and lists.lua (Colloquy 1.3x) files
in to executable perl code which is then evaluated. Colloquy 1.4 uses
a seperate LUA file for each user and list, which are located in the
users and lists directories in the Colloquy data directory. These files
are read one by one and evaluated in the same way.

This module compiles and execute the Colloquy data files in restricted
compartments using the L<Safe> module. Even so, this module should be
used with caution if you cannot gaurentee the integrity of the user and
list LUA files. The module will issue a warning complaining about world
writable permissions if $^W warnings.

=head1 EXPORTS

=head2 users

 my ($users_hashref,$lists_hashref) = users($colloquy_datadir);

Returns users and lists hash references, in that order.

=head2 lists

 my ($lists_hashref,$users_hashref) = lists($colloquy_datadir);

Returns lists and users hash references, in that order.

=head1 SEE ALSO

L<http://freshmeat.net/projects/colloquy-talker/>, L<Apache2::AuthColloquy>

=head1 VERSION

$Id$

=head1 AUTHOR

Nicola Worthington <nicolaw@cpan.org>

L<http://perlgirl.org.uk>

=head1 COPYRIGHT

Copyright 2005,2006 Nicola Worthington.

This software is licensed under The Apache Software License, Version 2.0.

L<http://www.apache.org/licenses/LICENSE-2.0>

=cut

__END__



